Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -477,6 +477,125 @@ void TextureCacheBase::BindTextures()
+    return sRGB * (sRGB * (sRGB * 0.305306011 + 0.682171111) + 0.012522878);
+  }
+
+  static u8 LinearToSRGB(float linear) {
+    return std::max(1.055f * std::pow(linear, 0.416666667f) - 0.055f, 0.f) * 256.f;
+  }
+  
+  struct Level {
+    u32 width;
+    u32 height;
+    u32 row_length;
+    const PixelRGBA* buffer;
+
+    PixelRGBAf Sample(u32 x, u32 y) const {
+      auto& p = buffer[x + y * row_length];
+      return {
+        SRGBToLinear(p[0]),
+        SRGBToLinear(p[1]),
+        SRGBToLinear(p[2]),
+        SRGBToLinear(p[3])
+      };
+    }
+
+    // Puts a downsampled image in dst. dst must be at least width*height*4
+    void Downsample(PixelRGBA* dst, const Level& mip_shape) const {
+      for (u32 i = 0; i < mip_shape.height; ++i) {
+        for (u32 j = 0; j < mip_shape.width; ++j) {
+          auto x = j * 2;
+          auto y = i * 2;
+          const std::array<PixelRGBAf, 4> samples = {
+            Sample(x, y),
+            Sample(x + 1, y),
+            Sample(x, y + 1),
+            Sample(x + 1, y + 1)
+          };
+          auto& dst_pixel = dst[j + i * mip_shape.row_length];
+          dst_pixel[0] = LinearToSRGB((samples[0][0] + samples[0][1] + samples[0][2] + samples[0][3]) * 0.25f);
+          dst_pixel[1] = LinearToSRGB((samples[1][0] + samples[1][1] + samples[1][2] + samples[1][3]) * 0.25f);
+          dst_pixel[2] = LinearToSRGB((samples[2][0] + samples[2][1] + samples[2][2] + samples[2][3]) * 0.25f);
+          dst_pixel[3] = LinearToSRGB((samples[3][0] + samples[3][1] + samples[3][2] + samples[3][3]) * 0.25f);
+
+        }
+      }
+    }
+
+    float AverageDiff(const PixelRGBA* other) const {
+      float average_diff = 0.f;
+      const auto* ptr1 = buffer;
+      const auto* ptr2 = other;
+      for (u32 i = 0; i < height; ++i) {
+        auto row1 = ptr1;
+        auto row2 = ptr2;
+        for (u32 j = 0; j < width; ++j, ++row1, ++row2) {
+          average_diff += std::abs((*row1)[0] - (*row2)[0]);
+          average_diff += std::abs((*row1)[1] - (*row2)[1]);
+          average_diff += std::abs((*row1)[2] - (*row2)[2]);
+          average_diff += std::abs((*row1)[3] - (*row2)[3]);
+        }
+        ptr1 += row_length;
+        ptr2 += row_length;
+      }
+
+      return average_diff / (width * height * 4) / 2.56f;
+    }
+  };
+
+public:
+  explicit ArbitraryMipmapDetector() {
+
+  }
+
+  void AddLevel(u32 width, u32 height, u32 row_length, const u8* buffer)
+  {
+    levels.push_back({ width, height, row_length, reinterpret_cast<const PixelRGBA*>(buffer) });
+  }
+
+  bool HasArbitraryMipmaps() const {
+    if (levels.size() < 2)
+      return false;
+
+    // This is the average per-pixel, per-channel difference in percent between what we
+    // expect a normal blurred mipmap to look like and what we actually received
+    constexpr auto THRESHOLD_PERCENT = 35.f;
+
+    static std::vector<PixelRGBA> downsampled;
```

Comment:
Is a static local *really* necessary here as opposed to being a private class variable?